File: backend/src/main/java/com/backend/petplace/domain/place/entity/Place.java
package com.backend.petplace.domain.place.entity;

import com.backend.petplace.global.entity.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Table(
    name = "place",
    indexes = {
        @Index(name = "idx_place_lat", columnList = "latitude"),
        @Index(name = "idx_place_lon", columnList = "longitude"),
        @Index(name = "idx_place_category2", columnList = "category2"),
        @Index(name = "idx_place_name", columnList = "name")
    }
)
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class Place extends BaseEntity {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @Column(nullable = false, length = 100, unique = true)
  private String uniqueKey;

  @Column(nullable = false, length = 200)
  private String name;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false, length = 40)
  private Category1Type category1;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false, length = 40)
  private Category2Type category2;

  @Column(length = 500)
  private String openingHours;

  @Column(length = 300)
  private String closedDays;

  private Boolean parking;

  private Boolean petAllowed;

  @Column(length = 300)
  private String petRestriction;

  @Column(length = 30)
  private String tel;

  @Column(length = 500)
  private String url;

  @Column(length = 10)
  private String postalCode;

  @Column(length = 300)
  private String address;

  @Column(nullable = false)
  private Double latitude;

  @Column(nullable = false)
  private Double longitude;

  @Column(length = 1000)
  private String rawDescription;

  @Builder.Default
  private Double averageRating = 0.0;

  @Builder.Default
  private Integer totalReviewCount = 0;

  public void updateReviewStats(int newRating) {
    double totalScore = this.averageRating * this.totalReviewCount;
    this.totalReviewCount++;
    this.averageRating = (totalScore + newRating) / this.totalReviewCount;
  }

  public void apply(String name, Category1Type c1, Category2Type c2, String openingHours,
      String closedDays, Boolean parking, Boolean petAllowed, String petRestriction,
      String tel, String url, String postalCode, String address, Double lat, Double lng,
      String rawDescription
  ) {
    this.name = name;
    this.category1 = c1;
    this.category2 = c2;
    this.openingHours = openingHours;
    this.closedDays = closedDays;
    this.parking = parking;
    this.petAllowed = petAllowed;
    this.petRestriction = petRestriction;
    this.tel = tel;
    this.url = url;
    this.postalCode = postalCode;
    this.address = address;
    this.latitude = lat;
    this.longitude = lng;
    this.rawDescription = rawDescription;
  }

  public Long getId() {
    return id;
  }

  public String getUniqueKey() {
    return uniqueKey;
  }

  public String getName() {
    return name;
  }

  public Category1Type getCategory1() {
    return category1;
  }

  public Category2Type getCategory2() {
    return category2;
  }

  public String getOpeningHours() {
    return openingHours;
  }

  public String getClosedDays() {
    return closedDays;
  }

  public Boolean getPetAllowed() {
    return petAllowed;
  }

  public Boolean getParking() {
    return parking;
  }

  public String getPetRestriction() {
    return petRestriction;
  }

  public String getTel() {
    return tel;
  }

  public String getUrl() {
    return url;
  }

  public String getPostalCode() {
    return postalCode;
  }

  public Double getLatitude() {
    return latitude;
  }

  public String getAddress() {
    return address;
  }

  public Double getLongitude() {
    return longitude;
  }

  public String getRawDescription() {
    return rawDescription;
  }

  public Double getAverageRating() {
    return averageRating;
  }

  public Integer getTotalReviewCount() {
    return totalReviewCount;
  }
}

File: backend/src/main/java/com/backend/petplace/domain/place/entity/Place.kt
package com.backend.petplace.domain.place.entity

import com.backend.petplace.global.entity.BaseEntity
import jakarta.persistence.*

@Entity
@Table(
    name = "place",
    indexes = [
        Index(name = "idx_place_lat", columnList = "latitude"),
        Index(name = "idx_place_lon", columnList = "longitude"),
        Index(name = "idx_place_category2", columnList = "category2"),
        Index(name = "idx_place_name", columnList = "name")
    ]
)
class Place(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @Column(nullable = false, length = 100, unique = true)
    var uniqueKey: String,

    @Column(nullable = false, length = 200)
    var name: String,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 40)
    var category1: Category1Type,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 40)
    var category2: Category2Type,

    @Column(length = 500)
    var openingHours: String? = null,

    @Column(length = 300)
    var closedDays: String? = null,

    var parking: Boolean? = null,

    var petAllowed: Boolean? = null,

    @Column(length = 300)
    var petRestriction: String? = null,

    @Column(length = 30)
    var tel: String? = null,

    @Column(length = 500)
    var url: String? = null,

    @Column(length = 10)
    var postalCode: String? = null,

    @Column(length = 300)
    var address: String? = null,

    @Column(nullable = false)
    var latitude: Double,

    @Column(nullable = false)
    var longitude: Double,

    @Column(length = 1000)
    var rawDescription: String? = null,

    // @Builder.Default 대체: 기본값 설정
    var averageRating: Double = 0.0,

    var totalReviewCount: Int = 0

) : BaseEntity() {

    fun updateReviewStats(newRating: Int) {
        val totalScore = this.averageRating * this.totalReviewCount
        this.totalReviewCount++
        this.averageRating = (totalScore + newRating) / this.totalReviewCount
    }

    fun apply(
        name: String,
        c1: Category1Type,
        c2: Category2Type,
        openingHours: String?,
        closedDays: String?,
        parking: Boolean?,
        petAllowed: Boolean?,
        petRestriction: String?,
        tel: String?,
        url: String?,
        postalCode: String?,
        address: String?,
        lat: Double,
        lng: Double,
        rawDescription: String?
    ) {
        this.name = name
        this.category1 = c1
        this.category2 = c2
        this.openingHours = openingHours
        this.closedDays = closedDays
        this.parking = parking
        this.petAllowed = petAllowed
        this.petRestriction = petRestriction
        this.tel = tel
        this.url = url
        this.postalCode = postalCode
        this.address = address
        this.latitude = lat
        this.longitude = lng
        this.rawDescription = rawDescription
    }
}
